### PR TITLE
lxd: introduce LXC to interface with lxc

### PR DESCRIPTION
**Pull request by cjp256**
To keep the size of the PR down, integration test will follow as a
separate PR, as it requires some additional helpers.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_providers/lxd/lxc.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+            raise LXDError(
+                brief=f"Failed to set config key {key!r} to {value!r} for instance {instance_name!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def delete(
+        self,
+        *,
+        instance_name: str,
+        force=False,
```
Comment (cjp256) [inline craft_providers/lxd/lxc.py:214]:
```suggestion
        force: bool = False,
```

File: craft_providers/lxd/lxc.py [hunk 2]
```diff
@@ -0,0 +1,972 @@
...
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def exec(
+        self,
+        *,
+        command: List[str],
+        instance_name: str,
+        cwd: str = "/root",
+        mode: str = "auto",
```
Comment (cjp256) [inline craft_providers/lxd/lxc.py:251]:
```suggestion
        mode: Optional[str] = None,
```

File: craft_providers/lxd/lxc.py [hunk 3]
```diff
@@ -0,0 +1,972 @@
...
+            "--project",
+            project,
+            "exec",
+            f"{remote}:{instance_name}",
+        ]
+
+        if cwd != "/root":
+            final_cmd.extend(["--cwd", cwd])
+
+        if mode != "auto":
```
Comment (cjp256) [inline craft_providers/lxd/lxc.py:284]:
```suggestion
        if mode is not None:
```

File: craft_providers/lxd/lxc.py [hunk 4]
```diff
@@ -0,0 +1,972 @@
...
+
+    def file_push(
+        self,
+        *,
+        instance_name: str,
+        source: pathlib.Path,
+        destination: pathlib.Path,
+        create_dirs: bool = False,
+        recursive: bool = False,
+        gid: int = -1,
```
Comment (cjp256) [inline craft_providers/lxd/lxc.py:350]:
```suggestion
        gid: Optional[int] = None,
```

File: craft_providers/lxd/lxc.py [hunk 5]
```diff
@@ -0,0 +1,972 @@
...
+        except subprocess.CalledProcessError as error:
+            raise LXDError(
+                brief=f"Failed to delete image {image!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def image_list(
+        self, *, project: str = "default", remote: str = "local"
+    ) -> List[Dict[str, Any]]:
+        """List instance_names.
```
Comment (cjp256) [inline craft_providers/lxd/lxc.py:571]:
```suggestion
        """List images.
```

**Review by sergiusens — APPROVED** [review]

File: craft_providers/lxd/lxc.py (hunk 3 above)
Comment (cmatsuoka) [reply craft_providers/lxd/lxc.py:284]: Wondering if this should be just `if mode`, for the case where `mode` is an empty string.

**Review by cmatsuoka — APPROVED** [review]

**Pull request merged**